File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/store/InMemoryOpsStore.java
package com.opspilot.store;

import com.opspilot.domain.Alert;
import com.opspilot.domain.AlertStatus;
import com.opspilot.domain.Exceptions;
import com.opspilot.domain.Incident;
import com.opspilot.domain.IncidentStatus;
import com.opspilot.domain.Runbook;
import com.opspilot.domain.Severity;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Porte de InMemoryStore (store/in-memory-store.ts) -- o próprio original usa
 * essa implementação em memória em testes; aqui ela é a implementação
 * default (o original usa SQLite em produção via SqliteOpsStore, fora do
 * escopo deste porte -- ver README).
 */
public class InMemoryOpsStore implements OpsStore {

    private final List<Alert> alerts = new ArrayList<>();
    private final List<Incident> incidents = new ArrayList<>();
    private final Map<String, Runbook> runbooks = new LinkedHashMap<>();

    @Override
    public List<Alert> getAlerts(AlertStatus status) {
        return alerts.stream()
                .filter(alert -> status == null || alert.status() == status)
                .collect(Collectors.toList());
    }

    @Override
    public List<Incident> getIncidents(IncidentStatus status) {
        return incidents.stream()
                .filter(incident -> status == null || incident.status() == status)
                .sorted((a, b) -> Long.compare(a.createdAt(), b.createdAt()))
                .collect(Collectors.toList());
    }

    @Override
    public Incident createIncident(String title, String service, Severity severity) {
        Incident incident = new Incident(
                "inc-" + System.currentTimeMillis() + "-" + UUID.randomUUID().toString().substring(0, 4),
                title, service, severity, IncidentStatus.OPEN, System.currentTimeMillis(), null, null);
        incidents.add(incident);
        return incident;
    }

    @Override
    public Incident resolveIncident(String id, String summary) {
        for (int i = 0; i < incidents.size(); i++) {
            if (incidents.get(i).id().equals(id)) {
                Incident resolved = incidents.get(i).resolved(System.currentTimeMillis(), summary);
                incidents.set(i, resolved);
                return resolved;
            }
        }
        throw new Exceptions.IncidentNotFoundException(id);
    }

    @Override
    public Runbook getRunbook(String service) {
        Runbook runbook = runbooks.get(service);
        if (runbook == null) {
            throw new Exceptions.RunbookNotFoundException(service);
        }
        return runbook;
    }

    @Override
    public void seedAlert(Alert alert) {
        alerts.add(alert);
    }

    @Override
    public void seedRunbook(Runbook runbook) {
        runbooks.put(runbook.service(), runbook);
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/store/inmemory_ops_store.go
package store

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"sync"
	"time"

	"opspilot/domain"
)

// InMemoryOpsStore é o porte de InMemoryStore (store/in-memory-store.ts) --
// o próprio original usa essa implementação em memória em testes; aqui ela é
// a implementação default (ver README). Protegida por mutex porque, ao
// contrário do event-loop single-threaded do Node, o net/http do Go atende
// requisições concorrentemente por padrão.
type InMemoryOpsStore struct {
	mu        sync.Mutex
	alerts    []domain.Alert
	incidents []domain.Incident
	runbooks  map[string]domain.Runbook
}

// NewInMemoryOpsStore cria um InMemoryOpsStore vazio, pronto para uso.
func NewInMemoryOpsStore() *InMemoryOpsStore {
	return &InMemoryOpsStore{
		runbooks: make(map[string]domain.Runbook),
	}
}

var _ OpsStore = (*InMemoryOpsStore)(nil)

func (s *InMemoryOpsStore) GetAlerts(status *domain.AlertStatus) []domain.Alert {
	s.mu.Lock()
	defer s.mu.Unlock()

	result := make([]domain.Alert, 0, len(s.alerts))
	for _, alert := range s.alerts {
		if status == nil || alert.Status == *status {
			result = append(result, alert)
		}
	}
	return result
}

func (s *InMemoryOpsStore) GetIncidents(status *domain.IncidentStatus) []domain.Incident {
	s.mu.Lock()
	defer s.mu.Unlock()

	result := make([]domain.Incident, 0, len(s.incidents))
	for _, incident := range s.incidents {
		if status == nil || incident.Status == *status {
			result = append(result, incident)
		}
	}
	// Já inserido em ordem de criação (createIncident faz append), igual ao
	// .sort por createdAt do original -- sem necessidade de reordenar.
	return result
}

func (s *InMemoryOpsStore) CreateIncident(title, service string, severity domain.Severity) domain.Incident {
	s.mu.Lock()
	defer s.mu.Unlock()

	incident := domain.Incident{
		ID:        fmt.Sprintf("inc-%d-%s", time.Now().UnixMilli(), randomHex(2)),
		Title:     title,
		Service:   service,
		Severity:  severity,
		Status:    domain.IncidentOpen,
		CreatedAt: time.Now().UnixMilli(),
	}
	s.incidents = append(s.incidents, incident)
	return incident
}

func (s *InMemoryOpsStore) ResolveIncident(id string, summary *string) (domain.Incident, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for i, incident := range s.incidents {
		if incident.ID == id {
			summaryValue := ""
			if summary != nil {
				summaryValue = *summary
			}
			resolved := incident.Resolved(time.Now().UnixMilli(), summaryValue)
			s.incidents[i] = resolved
			return resolved, nil
		}
	}
	return domain.Incident{}, &domain.IncidentNotFoundError{ID: id}
}

func (s *InMemoryOpsStore) GetRunbook(service string) (domain.Runbook, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	runbook, ok := s.runbooks[service]
	if !ok {
		return domain.Runbook{}, &domain.RunbookNotFoundError{Service: service}
	}
	return runbook, nil
}

func (s *InMemoryOpsStore) SeedAlert(alert domain.Alert) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.alerts = append(s.alerts, alert)
}

func (s *InMemoryOpsStore) SeedRunbook(runbook domain.Runbook) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.runbooks[runbook.Service] = runbook
}

// randomHex é o equivalente a UUID.randomUUID().toString().substring(0, 4) do
// porte Java: só precisa de baixa colisão para um id legível, não de um UUID
// completo.
func randomHex(n int) string {
	buf := make([]byte, n)
	_, _ = rand.Read(buf)
	return hex.EncodeToString(buf)
}
